ntoskrnl/io/iomgr/deviface.c

```diff
@@ -30,185 +31,435 @@ static PWCHAR BaseKeyString = L"\\Registry\\Machine\\System\\CurrentControlSet\\
+    }
+
+    /* Get reference string (starts after {GUID} and \) from link without prefix */
+    // TODO: Add some more sanity checks
+    ReferenceStringReal.Buffer = LinkNameNoPrefix.Buffer + ReferenceStringOffset / sizeof(WCHAR);
+    ReferenceStringReal.Length = LinkNameNoPrefix.Length - ReferenceStringOffset;
+    ReferenceStringReal.MaximumLength = ReferenceStringReal.Length;
+
+    /* Get GUID string (device class GUID in {} brackets) */
+    GuidStringReal.Buffer = LinkNameNoPrefix.Buffer + (ReferenceStringOffset - GUID_STRING_BYTES) / sizeof(WCHAR);
+    GuidStringReal.Length = GUID_STRING_BYTES;
+    GuidStringReal.MaximumLength = GuidStringReal.Length;
+
+    /* Create a temporary device string (looks like ACPI#PNP0501#1#) */
+    DeviceString.Buffer = LinkNameNoPrefix.Buffer;
+    DeviceString.Length = LinkNameNoPrefix.Length - ReferenceStringReal.Length;
+    DeviceString.MaximumLength = DeviceString.Length;
+
+    DPRINT("Device string is '%wZ'\n", &DeviceString);
+
+    if (MungedString != NULL)
+    {
+        /* Allocate a munged path string */
+        MungedStringReal.Length = 0;
+        MungedStringReal.MaximumLength = MungedPrefix.Length + DeviceString.Length + sizeof(UNICODE_NULL);
+        MungedStringReal.Buffer = ExAllocatePoolWithTag(PagedPool, MungedStringReal.MaximumLength, TAG_IO);
+        if (!MungedStringReal.Buffer)
+        {
+            DPRINT1("ExAllocatePoolWithTag() failed\n");
+            return STATUS_INSUFFICIENT_RESOURCES;
+        }
+
+        /* Fill it by munged prefix and our previously created device string */
+        Status = RtlAppendUnicodeStringToString(&MungedStringReal, &MungedPrefix);
+        NT_ASSERT(NT_SUCCESS(Status));
+
+        Status = RtlAppendUnicodeStringToString(&MungedStringReal, &DeviceString);
+        NT_ASSERT(NT_SUCCESS(Status));
+
+        DPRINT("Munged string is '%wZ'\n", &MungedStringReal);
+    }
+
+    DPRINT("GUID string is '%wZ'\n", &GuidStringReal);
+    DPRINT("Reference string is '%wZ'\n", &ReferenceStringReal);
+
+    /* Store received parts if the parameters are not null */
+    if (PrefixString != NULL)
+    {
+        if (RtlPrefixUnicodeString(&KernelModePrefix, SymbolicLinkName, FALSE))
+        {
+            PrefixString->Buffer = SymbolicLinkName->Buffer;
+            PrefixString->Length = KernelModePrefix.Length;
+            PrefixString->MaximumLength = PrefixString->Length;
+        }
+        else if (RtlPrefixUnicodeString(&UserModePrefix, SymbolicLinkName, FALSE))
+        {
+            PrefixString->Buffer = SymbolicLinkName->Buffer;
+            PrefixString->Length = UserModePrefix.Length;
+            PrefixString->MaximumLength = PrefixString->Length;
+        }
+
+        DPRINT("Prefix string is '%wZ'\n", PrefixString);
+    }
+
+    if (MungedString != NULL)
+        *MungedString = MungedStringReal;
     else
-        InstanceKeyRealP = &InstanceKeyReal;
+        RtlFreeUnicodeString(&MungedStringReal);
```

Comment:
If the output `MungedString` pointer is initially NULL, then the block above that initializes `MungedStringReal` is not taken, and therefore nothing has been allocated and therefore there is nothing to free in this else-case.

Reply:
Fixed, thanks.